File: src/main/java/at/xirado/bean/command/Command.java
package at.xirado.bean.command;

import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.internal.utils.Checks;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

import java.util.*;

public abstract class Command
{
    private final String name;
    private final String description;
    private final String usage;
    private final List<String> aliases = new ArrayList<>();
    private final List<Permission> requiredPermissions = new ArrayList<>();
    private final List<Permission> requiredBotPermissions = new ArrayList<>();
    private final List<Long> allowedGuilds = new ArrayList<>();
    private final EnumSet<CommandFlag> commandFlags;

    protected Command(String name, String description, String usage)
    {
        Checks.notNull(name, "name");
        Checks.notNull(description, "description");
        Checks.notNull(usage, "usage");

        this.name = name;
        this.description = description;
        this.usage = usage;
        commandFlags = EnumSet.noneOf(CommandFlag.class);
    }


    public void addAliases(@NotNull String... aliases)
    {
        Checks.notNull(aliases, "Aliases");
        this.aliases.addAll(Arrays.asList(aliases));
    }

    public void addRequiredPermissions(@NotNull Permission... permissions)
    {
        Checks.notNull(permissions, "Permissions");
        this.requiredPermissions.addAll(Arrays.asList(permissions));
    }

    public void addRequiredBotPermissions(@NotNull Permission... permissions)
    {
        Checks.notNull(permissions, "Permissions");
        this.requiredBotPermissions.addAll(Arrays.asList(permissions));
    }

    public void addAllowedGuilds(@NotNull Long... guildIDs)
    {
        Checks.notEmpty(guildIDs, "Guild Ids");
        this.allowedGuilds.addAll(Arrays.asList(guildIDs));
    }

    public void addCommandFlags(@NotNull CommandFlag... commandFlags)
    {
        Checks.notNull(commandFlags, "CommandFlags");
        this.commandFlags.addAll(Arrays.asList(commandFlags));
    }

    public String getName()
    {
        return name;
    }

    public String getDescription()
    {
        return description;
    }

    public String getUsage()
    {
        return usage;
    }


    public @NotNull List<String> getAliases()
    {
        return Collections.unmodifiableList(aliases);
    }

    public @NotNull List<Permission> getRequiredPermissions()
    {
        return Collections.unmodifiableList(requiredPermissions);
    }

    public @NotNull List<Permission> getRequiredBotPermissions()
    {
        return Collections.unmodifiableList(requiredBotPermissions);
    }

    public @NotNull List<Long> getAllowedGuilds()
    {
        return Collections.unmodifiableList(allowedGuilds);
    }

    public @NotNull Set<CommandFlag> getCommandFlags()
    {
        return Collections.unmodifiableSet(commandFlags);
    }

    public boolean hasCommandFlag(CommandFlag flag)
    {
        return commandFlags.contains(flag);
    }

    public boolean isAvailableIn(long guildID)
    {
        if (hasCommandFlag(CommandFlag.PRIVATE_COMMAND)) {
            return getAllowedGuilds().contains(guildID);
        } else {
            return true;
        }
    }

    public abstract void executeCommand(GuildMessageReceivedEvent event, CommandContext context);

    @NonNls
    @NotNull
    @Override
    public String toString() {
        return "Command{" +
                "name='" + name + '\'' +
                ", description='" + description + '\'' +
                ", usage='" + usage + '\'' +
                ", aliases=" + aliases +
                ", requiredPermissions=" + requiredPermissions +
                ", requiredBotPermissions=" + requiredBotPermissions +
                ", allowedGuilds=" + allowedGuilds +
                ", commandFlags=" + commandFlags +
                '}';
    }
}

File: src/main/java/at/xirado/bean/command/Command.kt
/*
 * Copyright 2024 Marcel Korzonek and contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package at.xirado.bean.command

import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.internal.utils.Checks
import java.util.*

abstract class Command protected constructor(name: String, description: String, usage: String) {

    val name: String
    val description: String
    val usage: String
    private val aliases: MutableList<String> = mutableListOf()
    private val requiredPermissions: MutableList<Permission> = mutableListOf()
    private val requiredBotPermissions: MutableList<Permission> = mutableListOf()
    private val allowedGuilds: MutableList<Long> = mutableListOf()
    private val commandFlags: EnumSet<CommandFlag> = EnumSet.noneOf(CommandFlag::class.java)

    init {
        this.name = name
        this.description = description
        this.usage = usage
    }

    fun addAliases(vararg aliases: String) {
        Checks.notNull(aliases, "Aliases")
        this.aliases.addAll(listOf(*aliases))
    }

    fun addRequiredPermissions(vararg permissions: Permission) {
        Checks.notNull(permissions, "Permissions")
        requiredPermissions.addAll(listOf(*permissions))
    }

    fun addRequiredBotPermissions(vararg permissions: Permission) {
        Checks.notNull(permissions, "Permissions")
        requiredBotPermissions.addAll(listOf(*permissions))
    }

    fun addAllowedGuilds(vararg guildIDs: Long) {
        Checks.notEmpty(guildIDs.toList(), "Guild Ids")
        allowedGuilds.addAll(guildIDs.toList())
    }

    fun addCommandFlags(vararg commandFlags: CommandFlag?) {
        Checks.notNull(commandFlags, "CommandFlags")
        this.commandFlags.addAll(listOf(*commandFlags))
    }

    fun getAliases(): List<String> {
        return Collections.unmodifiableList(aliases)
    }

    fun getRequiredPermissions(): List<Permission> {
        return Collections.unmodifiableList(requiredPermissions)
    }

    fun getRequiredBotPermissions(): List<Permission> {
        return Collections.unmodifiableList(requiredBotPermissions)
    }

    fun getAllowedGuilds(): List<Long> {
        return Collections.unmodifiableList(allowedGuilds)
    }

    fun getCommandFlags(): Set<CommandFlag> {
        return Collections.unmodifiableSet(commandFlags)
    }

    fun hasCommandFlag(flag: CommandFlag): Boolean {
        return commandFlags.contains(flag)
    }

    fun isAvailableIn(guildID: Long): Boolean {
        return if (hasCommandFlag(CommandFlag.PRIVATE_COMMAND)) getAllowedGuilds().contains(guildID) else true
    }

    abstract suspend fun executeCommand(event: MessageReceivedEvent, context: CommandContext)

    override fun toString(): String {
        return "Command{" +
                "name='" + name + '\'' +
                ", description='" + description + '\'' +
                ", usage='" + usage + '\'' +
                ", aliases=" + aliases +
                ", requiredPermissions=" + requiredPermissions +
                ", requiredBotPermissions=" + requiredBotPermissions +
                ", allowedGuilds=" + allowedGuilds +
                ", commandFlags=" + commandFlags +
                '}'
    }
}
